pkg/operator/startupmonitorreadiness/readiness_checks.go

```diff
@@ -0,0 +1,259 @@
+package startupmonitorreadiness
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/operatorclient"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/client-go/kubernetes"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/transport"
+)
+
+// KubeAPIReadinessChecker is a struct that holds necessary data
+// to perform a set of checks against a Kube API server to assess its health condition
+type KubeAPIReadinessChecker struct {
+	// configuration for authN/authZ against the server
+	// populated from kubeconfig and set by the startup monitor pod
+	restConfig *rest.Config
+
+	// client we use to perform HTTP checks
+	client *http.Client
+
+	// defined here for easier testing
+	baseRawURL string
+
+	kubeClient *kubernetes.Clientset
+}
+
+// TODO: uncomment when https://github.com/openshift/library-go/pull/1130 merges
+//var _ startupmonitor.ReadinessChecker = &KubeAPIReadinessChecker{}
+// var _ startupmonitor.WantsRestConfig = &KubeAPIReadinessChecker{}
+
+// New creates a new Kube API readiness checker
+func New() *KubeAPIReadinessChecker {
+	return &KubeAPIReadinessChecker{
+		baseRawURL: "https://localhost:6443",
+	}
+}
+
+// SetRestConfig called by startup monitor to provide a valid configuration for authN/authZ against Kube API server
+func (ch *KubeAPIReadinessChecker) SetRestConfig(restConfig *rest.Config) {
+	ch.restConfig = restConfig
```

Comment:
is this restConfig limited by the client-go parameters `QPS, burst, RateLimiter`?
if affirmative, should we override it?

Reply:
the default values QPS=5, Burst=10 should be enough, we are going to send just request per second.